### PR TITLE
Add support for canned cases on Hera + 3DVar bugfix

### LINKED ISSUE #79: Duplicate R2D2 and fix directories on Hera
Need to rsync files between Orion and Hera

Will likely also require code changes to app (define `R2D2_CONFIG`, etc.)

### PR DESCRIPTION
**Pull request by CoryMartin-NOAA**
Closes #79 

Now that R2D2_SHARED and fix files have been copied from Orion to Hera, this PR adds example YAML files to run the same canned cases on both Hera and Orion using `run_jedi_exe.py`.

The JSON modules are removed for Hera, because when they are included, it will validate YAML before running and will end the executable because of strings when it is expecting numbers (see #80).

A small change is added to the function to create the slurm submission script to limit the # of tasks on a node on Hera due to a bus error that was happening during runtime.

This also adds a bugfix for 3DVar where now that FV3-JEDI checks to ensure the increment variables are contained within the list of state variables (`ps` is added to the state variables list).

### CODE REVIEW

**Comment by CatherineThomas-NOAA** [comment]
I was able to successfully run the 3dhofx case on Hera.  It matches well with the case I previously ran on Orion.
orion_hera

I attempted to run the 3dvar and 4dhofx cases as well.  I do not have an Orion comparison for these cases yet (I will revisit after the holiday weekend).  The 3dvar case ran successfully but the analysis does not look correct to me:
t_inc
The 4dhofx failed outright:
``` 
[h21c05:224079:0:224079] Caught signal 7 (Bus error: nonexistent physical address)
srun: error: h21c05: task 29: Killed
srun: launch/slurm: _step_signal: Terminating StepId=33341568.0
slurmstepd: error: *** STEP 33341568.0 ON h21c04 CANCELLED AT 2022-07-01T18:28:01 ***
```

I don't have experience with these two cases, so I hope someone can let me know if this is expected behavior or not.

**Comment by CoryMartin-NOAA** [comment]
@CatherineThomas-NOAA that 3DVar case is concerning... I hope the increment write hasn't been screwed up by a recent code change. The 4DHofX error is the same one I saw for 3DHofX before reducing the number of tasks per node, so I bet it is just a memory usage issue.

**Comment by RussTreadon-NOAA** [comment]
Attempt to run 3dvar case on Orion.  Job did not run to completion.   `fv3jedi_var.x` aborted with the following traceback

```
Assertion failed: vars_ <= x1.variables() in diff, line 82 of /work2/noaa/da/cmartin/GDASApp/work/GDASApp/fv3-jedi/src/fv3jedi/Increment/Increment.cc caught in  (/work2/noaa/da/cmartin/GDASApp/work/GDASApp/oops/src/oops/runs/Run.cc +192 execute)
Exception: oops::Variational<FV3JEDI, UFO and IODA observations> terminating...
Exception stack:
Assertion failed: vars_ <= x1.variables() in diff, line 82 of /work2/noaa/da/cmartin/GDASApp/work/GDASApp/fv3-jedi/src/fv3jedi/Increment/Increment.cc
backtrace [1] stack has 14 addresses
(/work/noaa/da/jedipara/opt/modules/intel-2020.2/impi-2020.2/eckit/ecmwf-1.16.0/lib/libeckit.so+eckit::BackTrace::dump[abi:cxx11]())0x1b0
(/work/noaa/da/jedipara/opt/modules/intel-2020.2/impi-2020.2/eckit/ecmwf-1.16.0/lib/libeckit.so+eckit::Exception::Exception())0x17c
(/work/noaa/da/jedipara/opt/modules/intel-2020.2/impi-2020.2/eckit/ecmwf-1.16.0/lib/libeckit.so+eckit::AssertionFailed::AssertionFailed(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, eckit::CodeLocation const&))0x1f
(/work2/noaa/da/cmartin/GDASApp/work/GDASApp/build/bin/../lib64/libfv3jedi.so+fv3jedi::Increment::diff(fv3jedi::State const&, fv3jedi::State const&))0x26d
(/work2/noaa/stmp/rtreadon/gdas_single_test_3dvar/fv3jedi_var.x)

```

**Comment by CoryMartin-NOAA** [comment]
@RussTreadon-NOAA This sounds like what happened before my small 3DVar bugfix in this PR. Is `ps` in the YAML file from your failed test?

**Comment by RussTreadon-NOAA** [comment]
> @RussTreadon-NOAA This sounds like what happened before my small 3DVar bugfix in this PR. Is `ps` in the YAML file from your failed test?

Yes, several occurrences of `ps` are found in `gdas_variational.yaml`.

**Comment by CoryMartin-NOAA** [comment]
@RussTreadon-NOAA ok thanks, let me try again soon to see if I can reproduce the failure.

**Comment by RussTreadon-NOAA** [comment]
@CoryMartin-NOAA , I started over with a fresh clone of GDASApp and a new checkout of `feature/hera`.   

Now `./run_jedi_exe.py -c examples/run_jedi_exe/3dvar_orion.yaml` fails when attempting to stage `gdas.t06z.amsua_n19.2021122106.nc4`.   R2D2 looks in `/work2/noaa/da/cmartin/GDASApp/R2D2_SHARED/ncdiag/ob/oper_gdas/PT6H/2021-12-21/`.   The `2021-12-21` directory does not exist on Orion.   

We have a `2021-12-21` directory on Hera in `/scratch1/NCEPDEV/da/Cory.R.Martin/GDASApp/R2D2_SHARED/ncdiag/ob/oper_gdas/PT6H/`

**Comment by CoryMartin-NOAA** [comment]
@RussTreadon-NOAA that's my fault from being lazy and testing things... I have a new reference dataset for H(x) science evaluation but the app does not yet support it, so a symlink was used. The new experiment only has August 1-7 days for obs. I've changed the symlink back, so it should work again on Orion. Soon, I need to update the app to handle different 'ob_dump' values so that multiple reference datasets can be compared.

**Comment by RussTreadon-NOAA** [comment]
> @CoryMartin-NOAA , I started over with a fresh clone of GDASApp and a new checkout of `feature/hera`.
> 
> Now `./run_jedi_exe.py -c examples/run_jedi_exe/3dvar_orion.yaml` fails when attempting to stage `gdas.t06z.amsua_n19.2021122106.nc4`. R2D2 looks in `/work2/noaa/da/cmartin/GDASApp/R2D2_SHARED/ncdiag/ob/oper_gdas/PT6H/2021-12-21/`. The `2021-12-21` directory does not exist on Orion.
> 
> We have a `2021-12-21` directory on Hera in `/scratch1/NCEPDEV/da/Cory.R.Martin/GDASApp/R2D2_SHARED/ncdiag/ob/oper_gdas/PT6H/`

@CoryMartin-NOAA , let me pause here.  It looks like you are working on Orion.  Now the obs files are present.    `fv3jedi_var.x` fails with the previously mentioned `Assertion failed: vars_ <= xl.variables()` error.

**Comment by RussTreadon-NOAA** [comment]
@CoryMartin-NOAA .   OK.  It took a while but `3dvar_orion.yaml` runs to completion on Orion.   I get the similarly strange latlon increment output reported by @CatherineThomas-NOAA  

t_inc_orion

**Comment by CoryMartin-NOAA** [comment]
Thanks @RussTreadon-NOAA Hmm, this seems like either an interpolation issue, or an increment write issue. I'm leaning towards the former since I feel like I can see the 6 tiles fairly clearly.

**Comment by CoryMartin-NOAA** [comment]
A quick spot check of writing out the `fms restart` increments and looking at T for tile1, I don't see this banding. I wonder if it is a problem with the latlon interpolation? Perhaps something changed in interpolation in JEDI recently? I'll look tomorrow.

**Comment by CatherineThomas-NOAA** [comment]
I pulled @CoryMartin-NOAA's task-related change for 4dhofx and now the job runs to completion on Hera.  Diags are created and the plots with eva look similar to the previously posted 3D version.

**Comment by RussTreadon-NOAA** [comment]
fv3-jedi issue [#540](https://github.com/JCSDA-internal/fv3-jedi/issues/540) documents a problem in fv3-jedi `src/fv3jedi/IO/LatLon/fv3jedi_io_latlon_mod.f90`.   As currently coded array `array_with_halo` is allocated with negative value for the starting value of the index range.  

For the `3dvar_orion.yaml` test, `array_with_halo` is allocated as `array_with_halo(-2:51,-2,51,1:127)`.   With this setup `3dvar_orion.yaml` generates the following temperature increment field summed over all model layers
t_inc_bug

As a test the array indices for `array_with_halo` are changed to match those of the field loaded into the array.   With this change `3dvar_orion.yaml` generates the following temperature increment field summed over all model layers
t_inc_fix

`3dvar_orion.yaml` is configured to assimilate amsua_n19 and sondes.   The amsua_n19 orbit is evident in the increment field from the modified `fv3jedi_io_latlon_mod.f90`.

**Review by RussTreadon-NOAA — APPROVED** [review]
This PR primarily focuses on Hera.   Branch `feature/hera` was installed on Orion.   As part of this installation a bug in `fv3-jedi/src/fv3jedi/IO/LatLon/fv3jedi_io_latlon_mod.f90` was corrected.   The bug fix only impacts latlon output from the 3dvar test.    

The 3dhofx, 4dhofx, and 3dvar tests were run on Orion.  All tests ran to completion.   Visual inspection of latlon increments from the 3dvar test indicates acceptable behavior.   I do not have a control against which to compare output from the 3dhofx and 4dhofx tests.   @CatherineThomas-NOAA reports that Hera output from these tests is acceptable.
